src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
+        let skip_name_constraints = cert_is_self_issued(working_cert) && current_depth != 0;
+        if !skip_name_constraints {
+            accumulated_constraints.accumulate_san(working_cert_extensions)?;
+        }
+
+        accumulated_constraints.apply_and_accumulate(working_cert_extensions)?;
+
+        // Look in the store's root set to see if the working cert is listed.
+        // If it is, we've reached the end.
+        if self.store.contains(working_cert) {
+            return Ok(vec![working_cert.clone()]);
+        }
+
+        // Check that our current depth does not exceed our policy-configured
+        // max depth. We do this after the root set check, since the depth
+        // only measures the intermediate chain's length, not the root or leaf.
+        if current_depth > self.policy.max_chain_depth {
+            return Err(ValidationError::Other(
+                "chain construction exceeds max depth".into(),
+            ));
+        }
+
+        // Otherwise, we collect a list of potential issuers for this cert,
+        // and continue with the first that verifies.
+        let mut last_err: Option<ValidationError> = None;
+        for issuing_cert_candidate in self.potential_issuers(working_cert) {
+            // A candidate issuer is said to verify if it both
+            // signs for the working certificate and conforms to the
+            // policy.
+            let issuer_extensions = issuing_cert_candidate.extensions()?;
+            match self.policy.valid_issuer(
+                issuing_cert_candidate,
+                working_cert,
+                current_depth,
+                &issuer_extensions,
+            ) {
+                Ok(_) => {
+                    match self.build_chain_inner(
+                        issuing_cert_candidate,
+                        // NOTE(ww): According to RFC 5280, we should only
+                        // increase the chain depth when the certificate is **not**
+                        // self-issued. In practice however, implementations widely
+                        // ignore this requirement, and unconditionally increment
+                        // the depth with every chain member. We choose to do the same;
+                        // see `pathlen::self-issued-certs-pathlen` from x509-limbo
+                        // for the testcase we intentionally fail.
+                        //
+                        // Implementation note for someone looking to change this in the future:
+                        // care should be taken to avoid infinite recursion with self-signed
+                        // certificates in the intermediate set; changing this behavior will
+                        // also require a "is not self-signed" check on intermediate candidates.
+                        //
+                        // See https://gist.github.com/woodruffw/776153088e0df3fc2f0675c5e835f7b8
+                        // for an example of this change.
```

Comment:
Here's my explanation of the self-signed/issued behavioral change. I've also linked a diff with a negative diff of the changes here, for someone in the future to build off of.